### PR TITLE
Explore: Virtualizes typeahead list

### LINKED ISSUE #13715: Explore: typeahead is slow when having lots of suggestions
On prometheus installations with 1000+ metrics the typeahead feels sluggish. 

Evaluate approaches to not render all suggestions.

- https://github.com/bvaughn/react-window
- https://github.com/bvaughn/react-virtualized

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Performance for typeahead list in Explore is bad when there are a lot of suggestions. This PR makes the performance better and constant by using virtualization.

**Before**
image

**After**
image

**Which issue(s) this PR fixes**:
Closes #13715 

**Special notes for your reviewer**:
Checkout branch and test.

### CODE REVIEW

**Comment by torkelo** [comment]
@hugohaggmark we are already using a virtualization library, react-virtualized so not sure we need react-window

**Comment by torkelo** [comment]
ah, reading up on react-window, looks like same author and smaller more focused library, interesting.

File: public/app/features/explore/TypeaheadItem.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  isSelected: boolean;
+  item: CompletionItem;
+  onClickItem: (suggestion: CompletionItem) => void;
+  prefix?: string;
+  style: any;
+}
+
+const getStyles = (theme: GrafanaTheme) => ({
+  typeaheadItem: css`
+    height: auto;
```
Comment (dprokop) [inline public/app/features/explore/TypeaheadItem.tsx]: FYI for easier debugging it's wise to add `label: component-name;` to css. Thanks to that generated class name will have `component-name` in it and it's easier to reason where the styles come from.

**Comment by hugohaggmark** [comment]
Yes, the react-window seems to be the future.

File: public/app/features/explore/TypeaheadItem.tsx (hunk 1 above)
Comment (hugohaggmark) [reply public/app/features/explore/TypeaheadItem.tsx]: Thanks, I'm still not a huge fan of the 
```typescript 
css`...`
```
syntax, it takes some getting used to.

**Comment by hugohaggmark** [comment]
Found a bug in Prometheus and the history/metrics, behaves strange when grouped like this
image

Every group becomes a virtualized list and there is no way to scroll down to functions using `arrow down`, the cursor is stuck on the last item in the history list.

**Comment by davkal** [comment]
I cant see the function help texts:
Screenshot 2019-04-03 at 16 37 31
Am I doing something wrong? Using `yarn start:hot`.

**Comment by davkal** [comment]
The first render speedup is astonishing. Testing with with 1700 metrics first mount latency goes from 600ms to 20ms. Scrolling down produces perhaps too many lazily loaded mounts (visibly empty list items), is there something we can tune?

**Comment by hugohaggmark** [comment]
> I cant see the function help texts:
> Screenshot 2019-04-03 at 16 37 31
> Am I doing something wrong? Using `yarn start:hot`.

I've seen the same issue with the new `emotion` theming when I tried the light theme. I ended up with running `env GRAFANA_THEME=light yarn start:hot` and choosing the `light` theme specifically in the Grafana settings. Maybe there's some new limitation to our theming combined with emotion. Any thoughts @dprokop ?

**Comment by hugohaggmark** [comment]
> The first render speedup is astonishing. Testing with with 1700 metrics first mount latency goes from 600ms to 20ms. Scrolling down produces perhaps too many lazily loaded mounts (visibly empty list items), is there something we can tune?

`react-window` is completely new to me but I've seen that there are some tweaks like `overscanCount` (The number of items (rows or columns) to render outside of the visible area.) and `estimatedItemSize` (Estimated size of a item in the direction being windowed.). There's also the code I wrote, that has not been optimized for performance.

Feels kind of nit-pick for me but if you feel strongly about this I could do another iteration.

**Comment by torkelo** [comment]
For theming to work the Grafana theme selected in user preferences must match the GRAFANA_THEME env var. As the ENV var only controls what sass theme gets compiled (not caring about the Grafana theme profile setting). And any frontend code that cares about style will not know about GRAFANA_THEME env var but will only check the profile setting.

**Comment by davkal** [comment]
I've tried it with `yarn run themes:generate` and then `yarn dev`. I'm still getting non-readable item descriptions:

Screenshot 2019-04-04 at 17 38 45
(The one readable is the one selected via keys. When scrolling with mouse, I wont see any descriptions.)

Re virtual tuning, I'm not fussed, we can leave this as is.

**Comment by torkelo** [comment]
@davkal the lack of description I see in master as well, it's only shown on the first function in the list.

But the strange height of the function items I am also seeing.

**Comment by torkelo** [comment]
Ah, now I have tested master and this branch again. The description in master is only shown when you select the function using the keyboard (down arrow), not on hover. In master it is display:none otherwise so it does not take any height but looks like here it's visibility: hidden (or the effect of it) as its hidden when the focus is not selected but still takes the same height as the description would take.

**Comment by hugohaggmark** [comment]
This is how react-window works, you have to be specific with item heights because that calculates how many of the items are shown. Maybe it's better if we always show the function documentations?

**Comment by torkelo** [comment]
An idea is to use a tooltip on the right side as in the prometheus dashboard query editor. Not sure about the blue background but otherwise looks good. 

Also would be nice if these descriptions show on mouse hover like in the dashboard query editor and not only on keyboard focus like today. 

Screen Shot 2019-04-05 at 07 13 19

**Comment by davkal** [comment]
As agreed just now in meeting:

- solution have consistent heights of list elements
- only need to display details of the selected element

File: public/app/features/explore/Typeahead.tsx [hunk 2]
```diff
@@ -1,107 +1,190 @@
...
-    const label = item.label || '';
-    return (
-      <li ref={this.getRef} className={className} onClick={this.onClick}>
-        <Highlighter textToHighlight={label} searchWords={[prefix]} highlightClassName="typeahead-match" />
-        {item.documentation && isSelected ? <div className="typeahead-item-hint">{item.documentation}</div> : null}
-      </li>
-    );
-  }
-}
+  calculateLongestLabel = (allItems: CompletionItem[]): string => {
```
Comment (davkal) [inline public/app/features/explore/Typeahead.tsx]: Could extract, has no `this`.

File: public/app/features/explore/Typeahead.tsx [hunk 3]
```diff
@@ -1,107 +1,190 @@
...
+  };
 
-interface TypeaheadGroupProps {
-  items: CompletionItem[];
-  label: string;
-  onClickItem: (suggestion: CompletionItem) => void;
-  selected: CompletionItem;
-  prefix?: string;
-}
+  calculateListSizes = (theme: GrafanaTheme, allItems: CompletionItem[], longestLabel: string) => {
```
Comment (davkal) [inline public/app/features/explore/Typeahead.tsx]: Could extract, has no `this`.

File: public/app/features/explore/TypeaheadInfo.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    const documentation = item && item.documentation ? item.documentation : '';
+    const styles = this.getStyles(theme, visible);
+
+    return (
+      <div
+        className={cx([styles.typeaheadItem])}
+        style={{ position: 'fixed', left: `${width + 164}px`, width: '250px', height: `${height + 1}px` }}
+      >
+        <b>{label}</b>
+        <hr />
```
Comment (davkal) [inline public/app/features/explore/TypeaheadInfo.tsx]: Wild!

**Review by davkal — APPROVED** [review]
Looks great and works well.

I was surprised to see so much calculation being necessary. 

On Loki, there the info box slightly overlaps.
Screenshot 2019-04-08 at 18 03 23


Bit sad that history metadata is short and has to be rendered in the box now, but let's roll with this.

File: public/app/features/explore/Typeahead.tsx (hunk 2 above)
Comment (hugohaggmark) [reply public/app/features/explore/Typeahead.tsx]: Yes!, extracted to separate file.

File: public/app/features/explore/Typeahead.tsx (hunk 3 above)
Comment (hugohaggmark) [reply public/app/features/explore/Typeahead.tsx]: Yes!, extracted to separate file.

File: public/app/features/explore/TypeaheadInfo.tsx (hunk 4 above)
Comment (hugohaggmark) [reply public/app/features/explore/TypeaheadInfo.tsx]: Changed so the TypeaheadInfo's position is calculated dynamically instead.

**Pull request merged**